### ISSUE TITLE
Local function in struct accessing this (error CS1673)

### ISSUE THREAD

**Issue by KalleOlaviNiemitalo** [dotnet/csharpstandard#926]
**Describe the bug**

The standard should forbid a local function in a member of a `struct` type from accessing `this` or instance members of `this`.  I don't see where the C# 7.x draft does that.

**Example**

The following should not be allowed.  [SharpLab](https://sharplab.io/#v2:CYLg1APgzgLgTgVwMYwAQGVUG8CwAoVQ1ASwDs1iBufI1GogAQBZUBZACgEpt7bDMAclx4E+YhgHZUMABbEolQgHolqAKZw4AezioAwugCMANgkBmXnwC+l2raJk0AeWG5RY2pJKKV6zTv0jUwt3a0sbPAigA===)

```csharp
struct S {
    int i;
    
    void M() {
        S N() {
            return this;  // error CS1673
        }
        
        int O() {
            return i; // error CS1673
        }
    }
}
```

**Expected behavior**

In [§13.6.4 (Local function declarations)](https://github.com/dotnet/csharpstandard/blob/d538391e2615696c873282af2dec65ed6493fb8b/standard/statements.md#1364-local-function-declarations), declare that a local function in a member of a `struct` shall not access `this` or instance members of `this`.  Static members are okay though.

Add the constraint in or near this paragraph, which talks about variables captured by local functions: <https://github.com/dotnet/csharpstandard/blob/d538391e2615696c873282af2dec65ed6493fb8b/standard/statements.md#L518>

**Additional context**

This is a new problem in C# 7, as the equivalent thing for anonymous delegates and lambda expressions is disallowed in [§12.19.3 (Anonymous function bodies)](https://github.com/dotnet/csharpstandard/blob/d538391e2615696c873282af2dec65ed6493fb8b/standard/expressions.md#12193-anonymous-function-bodies): <https://github.com/dotnet/csharpstandard/blob/d538391e2615696c873282af2dec65ed6493fb8b/standard/expressions.md?plain=1#L4978>

**Comment by jskeet** [comment]
Will look at this in detail on Monday - but just to set expectations, we're getting to the stage where *relatively* minor problems (even when we accept that they're genuine problems) may well be punted to the next version.

*Please* keep reporting stuff - I just want to set practical expectations about having to actually ship at some point.

**Comment by jskeet** [comment]
(And where we *can* fix stuff before C# 7 ships, we will, of course...)

**Comment by KalleOlaviNiemitalo** [comment]
Is there a risk that ECMA members will vote against the standard if it has too many known errors?  [ECMA By-laws](https://www.ecma-international.org/policies/by-laws/) §9.1 _"The adoption of such documents for publication by the Association shall require approval by at least two thirds of all the ordinary members."_ but I imagine some of the ordinary members don't care about C# and might approve without reading.

**Comment by KalleOlaviNiemitalo** [comment]
A surprising interaction with `nameof`:

```csharp
struct S {
    int i;
    
    void M() {
        string N() {
            return nameof(this);  // error CS1673
        }

        string O() {
            return nameof(this.i);  // error CS1673
        }

        string P() {
            return nameof(i);  // OK, returns "i"
        }
    }
}
```

**Comment by jskeet** [comment]
> Is there a risk that ECMA members will vote against the standard if it has too many known errors?

I'd be very surprised if that happens. I think there's a general recognition that *any* specification of this sort of size will have some issues. Better that we have them logged so we can improve things over time than that we don't know about them in the first place :)

**Comment by KalleOlaviNiemitalo** [comment]
Add into 13.6.4 (Local function declarations), above "A local function may be called":

> When the type of `this` is a struct type, it is a compile-time error for the body of a local function to access `this`. This is true whether the access is explicit (as in `this.x`) or implicit (as in `x` where `x` is an instance member of the struct). This rule simply prohibits such access and does not affect whether member lookup results in a member of the struct. *Note:* this rule mirrors the one for anonymous functions in 12.19.3. *end note*

Or:

> Access to `this` in a local function is restricted in the same way as in an anonymous function body (12.19.3).

To be checked:

- `goto` into or out of a local function

  *Tracked as <https://github.com/dotnet/csharpstandard/issues/930>*

- `nameof(this)` in a default parameter or in an attribute argument (thus not in the function body) in a local or anonymous function in a struct

  *Tracked as <https://github.com/dotnet/csharpstandard/issues/931>*

**Comment by KalleOlaviNiemitalo** [comment]
*Tracked as <https://github.com/dotnet/csharpstandard/issues/931>*

```csharp

using System;
class DAttribute : Attribute {
    public DAttribute(string s) {}
}
struct S {
    int j;
    void LeA() {
        // OK
        Action<int> a = ([D(nameof(this.j))] int i) => {};
    }
    void LeD() {
        // error CS1065, default value not allowed
        Action<string> action = (string s = nameof(this.j)) => {};
    }
    void LfA() {
        // OK
        void F([D(nameof(this.j))] int i) {}
    }
    void LfD() {
        // error CS1673, cannot access this in local function
        void F(string s = nameof(this.j)) {}
    }
}
```

That seems pretty inconsistent. Why is `nameof(this.j)` allowed in an attribute argument in the parameter list, but not in a default parameter value?  Is that what the standard should specify?

**Comment by jskeet** [comment]
I agree, that does seem inconsistent - post-C#-7 we can work on thinking about what we want the spec to say, and it's possible that we'll write it so that the current Roslyn behavior is a bug (that could then be fixed, it it's a matter of making the default parameter value case a valid one).

**Comment by KalleOlaviNiemitalo** [comment]
*Tracked as <https://github.com/dotnet/csharpstandard/issues/931>*

Expression trees seem to allow neither attributes nor default values, so the behaviour is more consistent for them:

```csharp
using System;
using System.Linq.Expressions;
class DAttribute : Attribute {
    public DAttribute(string s) {}
}
struct S {
    int j;
    void ExA() {
        // error CS8972, attribute not allowed in expression tree
        Expression<Func<string, string>> e = ([D(nameof(this.j))] string s) => "";
    }
    void ExD() {
        // error CS1065, default value not allowed in expression tree
        Expression<Func<string, string>> e = (string s = nameof(this.j)) => "";
    }
    void ExR() {
        // error CS1673, lambda expression in struct cannot access this
        Expression<Func<string, string>> e = (string s) => nameof(this.j);
    }
}
```

**Comment by jskeet** [comment]
As an aside: I'm hoping to get to the use of `this` within a local function, but I doubt that I'll make any changes for `nameof(this.i)` soon. I'll probably create a new issue for the `nameof` bits.

**Comment by jskeet** [comment]
For `goto` statements, I think we're already covered for "into":

> A label can be referenced from `goto` statements ([§13.10.4](statements.md#13104-the-goto-statement)) within the scope of the label.

> *Note*: This means that `goto` statements can transfer control within blocks and out of blocks, but never into blocks. *end note*

A local function declaring a label would have to have a block body, so I don't think you'd be able to goto a label within it.

Worth checking whether goto *outside* the local function (or anonymous function) is actually prohibited in the spec though. (It clearly shouldn't be allowed.)

```csharp
class Program
{
    static void Main()
    {
        LocalFunction();
    testlabel:

        Action action = () =>
        {
            if (DateTime.UtcNow.Second > 10)
            {
                goto testlabel;
            }
        };

        if (DateTime.UtcNow.Second < 5)
        {
            goto testlabel;
        }

        void LocalFunction()
        {
            if (DateTime.UtcNow.Second > 10)
            {
                goto testlabel;
            }
        }
    }
}
```

Same error for both: CS0159 No such label 'testlabel' within the scope of the goto statement

(Interestingly enough, VS *does* autocomplete testlabel, and renaming testlabel renames the targets of the goto...)

I suspect we *might* be able to fix this in 7.3, which currently states:

> Each *block* or *switch_block* creates a separate declaration space for labels. Names are introduced into this declaration space through *labeled_statement*s, and the names are referenced through *goto_statement*s. The ***label declaration space*** of a block includes any nested blocks. Thus, within a nested block it is not possible to declare a label with the same name as a label in an enclosing block.

Although declaring `testlabel` again within `LocalFunction` gives a different error:

> CS0158 The label 'testlabel' shadows another label by the same name in a contained scope

I'll move all this to a separate issue...

**Comment by jskeet** [comment]
Okay, I've copied the relevant bits into #930 and #931 - I'll leave them here at least for the moment as well, but I'll be happy to close this by merging #929.
